File: resources/android/CalloutRenderer.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import com.nativephp.mobile.ui.nativerender.NativeUITree
import com.nativephp.plugins.native_ui.NativeUITheme

sealed interface CalloutRendererEvent {
    data class Press(val callbackId: Int, val nodeId: Int) : CalloutRendererEvent
}

data class CalloutRendererConfiguration(
    val nodeId: Int,
    val message: String,
    val tone: CalloutTone,
    val actionLabel: String,
    val accessibilityLabel: String,
    val accessibilityHint: String,
    val callbackId: Int,
) {
    constructor(node: NativeUINode) : this(
        nodeId = node.id,
        message = node.props.getString("message"),
        tone = CalloutTone.fromWire(node.props.getString("tone", "info")),
        actionLabel = node.props.getString("action_label"),
        accessibilityLabel = node.props.getString("a11y_label"),
        accessibilityHint = node.props.getString("a11y_hint"),
        callbackId = node.onPress,
    )

    val hasAction: Boolean
        get() = actionLabel.isNotEmpty() && callbackId != 0

    val resolvedAccessibilityDescription: String
        get() = listOf(
            firstlightCalloutAccessibilityLabel(message, tone, accessibilityLabel),
            accessibilityHint,
        ).filter(String::isNotEmpty).joinToString(". ")

    fun pressEvent(): CalloutRendererEvent.Press? = if (hasAction) {
        CalloutRendererEvent.Press(callbackId = callbackId, nodeId = nodeId)
    } else {
        null
    }
}

class CalloutRendererState(node: NativeUINode) {
    var configuration by mutableStateOf(CalloutRendererConfiguration(node))
        private set

    fun serverPublished(tree: NativeUITree): Boolean {
        val node = tree.root.findCalloutNode(configuration.nodeId) ?: return false
        val published = CalloutRendererConfiguration(node)
        if (published == configuration) return false

        configuration = published
        return true
    }
}

object CalloutRenderer {
    @Composable
    fun Render(node: NativeUINode, modifier: Modifier) {
        val state = remember(node.id) { CalloutRendererState(node) }
        val publishedTree = NativeUIBridge.currentTree.value
        LaunchedEffect(publishedTree) {
            publishedTree?.let(state::serverPublished)
        }

        val configuration = state.configuration
        FirstlightCalloutControl(
            configuration = configuration,
            tokens = if (isSystemInDarkTheme()) NativeUITheme.dark else NativeUITheme.light,
            onPress = {
                configuration.pressEvent()?.let { event ->
                    NativeUIBridge.sendPressEvent(event.callbackId, event.nodeId)
                }
            },
            modifier = modifier,
        )
    }
}

private fun NativeUINode.findCalloutNode(id: Int): NativeUINode? {
    if (this.id == id) return this
    children.forEach { child -> child.findCalloutNode(id)?.let { return it } }
    return null
}

File: resources/ios/CalloutRenderer.swift
import Combine
import SwiftUI

enum CalloutRendererEvent: Equatable {
    case press(callbackID: Int, nodeID: Int)
}

struct CalloutRendererConfiguration: Equatable {
    let nodeID: Int
    let message: String
    let tone: CalloutTone
    let actionLabel: String
    let accessibilityLabel: String
    let accessibilityHint: String
    let callbackID: Int

    init(node: NativeUINode) {
        nodeID = node.id
        message = node.props.getString("message")
        tone = CalloutTone(
            rawValue: node.props.getString("tone", default: "info")
        ) ?? .info
        actionLabel = node.props.getString("action_label")
        accessibilityLabel = node.props.getString("a11y_label")
        accessibilityHint = node.props.getString("a11y_hint")
        callbackID = node.onPress
    }

    var hasAction: Bool {
        !actionLabel.isEmpty && callbackID != 0
    }

    var resolvedAccessibilityLabel: String {
        firstlightCalloutAccessibilityLabel(
            message: message,
            tone: tone,
            explicit: accessibilityLabel
        )
    }

    func pressEvent() -> CalloutRendererEvent? {
        guard hasAction else { return nil }

        return .press(callbackID: callbackID, nodeID: nodeID)
    }
}

struct CalloutRendererState {
    private(set) var configuration: CalloutRendererConfiguration

    init(node: NativeUINode) {
        configuration = CalloutRendererConfiguration(node: node)
    }

    @discardableResult
    mutating func serverPublished(_ tree: NativeUITree) -> Bool {
        guard let node = calloutNode(id: configuration.nodeID, in: tree.root) else {
            return false
        }

        let published = CalloutRendererConfiguration(node: node)
        guard published != configuration else { return false }

        configuration = published
        return true
    }
}

struct CalloutRenderer: View {
    let node: NativeUINode

    @ObservedObject private var themeStore = NativeUITheme.shared
    @Environment(\.colorScheme) private var colorScheme
    @State private var state: CalloutRendererState

    init(node: NativeUINode) {
        self.node = node
        _state = State(initialValue: CalloutRendererState(node: node))
    }

    var body: some View {
        let configuration = state.configuration

        FirstlightCalloutControl(
            configuration: configuration,
            tokens: themeStore.resolve(for: colorScheme),
            onPress: {
                guard case let .press(callbackID, nodeID) = configuration.pressEvent() else {
                    return
                }

                NativeUIBridge.sendPressEvent(callbackID, nodeId: nodeID)
            }
        )
        .onReceive(NativeUIBridge.shared.$currentTree.compactMap { $0 }) { tree in
            state.serverPublished(tree)
        }
    }
}

private func calloutNode(id: Int, in node: NativeUINode) -> NativeUINode? {
    if node.id == id { return node }

    for child in node.children {
        if let found = calloutNode(id: id, in: child) {
            return found
        }
    }

    return nil
}
